### PR TITLE
deploy with address

### PR DESCRIPTION
**Pull request by sunce86**
This PR adds possibility to deploy new contracts to the network using the custom contract address instead of using the internal counter for automatic address generation.

### CODE REVIEW

**Review by nlordell — APPROVED** [review]
LGTM

File: ethcontract-mock/src/details/mod.rs [hunk 1]
```diff
@@ -93,6 +93,13 @@ impl MockTransport {
         address
     }
 
+    /// Deploys a new contract with the given ABI and address
+    pub fn deploy_with_address(&self, abi: &Abi, address: Address) {
+        let mut state = self.state.lock().unwrap();
+
+        state.contracts.insert(address, Contract::new(address, abi));
```
Comment (nlordell) [inline ethcontract-mock/src/details/mod.rs]: Maybe we can assert that we aren't replacing an existing contract?

Comment (sunce86) [reply ethcontract-mock/src/details/mod.rs]: Do you mean adding a simple assert like this:
`assert!(!state.contracts.contains_key(&address));`

**Review by vkgnosis — APPROVED** [review]

File: ethcontract-mock/src/details/mod.rs (hunk 1 above)
Comment (nlordell) [reply ethcontract-mock/src/details/mod.rs]: Also, `insert` returns the previous value, so you can also:
```suggestion
        assert!(
            state.contracts.insert(address, Contract::new(address, abi)).is_none(),
            "replacing contract at address {address:?}",
        );
```

Comment (sunce86) [reply ethcontract-mock/src/details/mod.rs]: Added.

**Pull request merged**